### PR TITLE
Fix duplicate check with different mac format

### PR DESCRIPTION
**Pull request by fossedihelm**
A macAddress can be specified with different separators:
 - ":" e.g. "02:00:00:00:ff:ff"
 - "-" e.g. "02-00-00-00-ff-ff"
 - "." e.g. "0200.0000.ffff"

Currently, the duplicate mac check works only if the two compared mac address use the same separator.
The duplicate check is made looking at the existence of key of the macMap. The keys of the mapMap are the macAddress string.

With this commit, a `macKey` struct is introduced. The latter stores the normalized string of the macAddress and can be used as unique key and identifier of a macAddress.

**What this PR does / why we need it**:

**Special notes for your reviewer**:
Fixes https://issues.redhat.com/browse/CNV-30974

**Release note**:

```release-note
Fix duplicate check with different mac address separator
```

### CODE REVIEW

**Review by maiqueb — CHANGES_REQUESTED** [review]
This PR is required but IMO it doesn't do enough since there are other supported types of format for MAC addresses:
- canonical (`-` separator, which your PR adds support for)
- hexstring (`:` separator, the only thing currently accepted by kubemacpool)
- dot notation (`.` separator, in 4 char segments - e.g. `0000.5e00.5301`)
- hexadecimal 

Maybe the correct thing to do is to create a MAC address - `net.HardwareAddr` - using `net.ParseMAC` and then get the string from it (whatever that is) ?

**Comment by fossedihelm** [comment]
> Maybe the correct thing to do is to create a MAC address - `net.HardwareAddr` - using `net.ParseMAC` and then get the string from it (whatever that is) ?

@maiqueb That's a great idea! IMHO we can also skip the `err` from ParseMAC function, since it has already been validated. WDYT?

**Comment by maiqueb** [comment]
> > Maybe the correct thing to do is to create a MAC address - `net.HardwareAddr` - using `net.ParseMAC` and then get the string from it (whatever that is) ?
> 
> @maiqueb That's a great idea! IMHO we can also skip the `err` from ParseMAC function, since it has already been validated. WDYT?

Fine with me. I mean, there's nothing wrong in being redundant - checking against nil won't take long - but if you're sure that was already validated, all good. I just find it more readable / comforting to see the (ugly) golang pattern of `if err != nil ; return err` . It's up to the maintainer.

**Review by maiqueb — APPROVED** [review]
Looks good !

**Comment by fossedihelm** [comment]
> Fine with me. I mean, there's nothing wrong in being redundant - checking against nil won't take long - but if you're sure that was already validated, all good. I just find it more readable / comforting to see the (ugly) golang pattern of `if err != nil ; return err` . It's up to the maintainer.

Yes, I generally agree! In this  specific case, handling the error means to create something like this:
```
func NewMacKey(macAddressDashes string) (macKey, error) {
	macAddress, err := net.ParseMAC(macAddressDashes)
        if err != nil{
            return macKey{}, err
        }
	return macKey{normalizedMacAddress: macAddress.String()}, nil
}
```
This is not a problem at all, but this will not allow us to create inline entry in the macMap.
And the latter is very useful in tests

**Comment by maiqueb** [comment]
> > Fine with me. I mean, there's nothing wrong in being redundant - checking against nil won't take long - but if you're sure that was already validated, all good. I just find it more readable / comforting to see the (ugly) golang pattern of `if err != nil ; return err` . It's up to the maintainer.
> 
> Yes, I generally agree! In this specific case, handling the error means to create something like this:
> 
> ```
> func NewMacKey(macAddressDashes string) (macKey, error) {
> 	macAddress, err := net.ParseMAC(macAddressDashes)
>         if err != nil{
>             return macKey{}, err
>         }
> 	return macKey{normalizedMacAddress: macAddress.String()}, nil
> }
> ```
> 
> This is not a problem at all, but this will not allow us to create inline entry in the macMap. And the latter is very useful in tests, in addition to the fact that we should clearly always handle the return error.

Maybe it's best to add a comment to the `NewMacKey` func indicating it's inputs **must** be already validated MAC addresses ? And indicate all it does is parse from N mac address formats to "our" definition of canonical format.

This is just a nit. Consider adding it if someone else asks you to change the code.

**Comment by fossedihelm** [comment]
> Maybe it's best to add a comment to the `NewMacKey` func indicating it's inputs **must** be already validated MAC addresses ? And indicate all it does is parse from N mac address formats to "our" definition of canonical format.
> 
> This is just a nit. Consider adding it if someone else asks you to change the code.

I put this just before the ParseMAC func:
```
// we can ignore the error here since the string value has already been validated
```
but yes, I can insert the description above the func! Thanks, will do!

**Review by maiqueb — APPROVED** [review]

File: pkg/pool-manager/pool.go
Comment (RamLavi) [inline pkg/pool-manager/pool.go:82]: I think this one deserves a separate file. Please move this to a new file `macKey.go`, and add the appropriate `macKey_tests.go` with unit tests for it (nothing fancy, I think this is enough:
```
var _ = Describe("MacKey", func() {
	DescribeTable("String should return the normalized MAC address",
		func(input string, expected string) {
			macKey := macKey{normalizedMacAddress: input}
			Expect(macKey.String()).To(Equal(expected))
		},
		Entry("colon-separated input", "00:00:5e:00:53:01", "00:00:5e:00:53:01"),
		Entry("hyphen-separated input", "00-00-5e-00-53-01", "00:00:5e:00:53:01"),
		Entry("period-separated input", "0000.5e00.5301", "00:00:5e:00:53:01"),
	)
})
```
)

File: pkg/pool-manager/virtualmachine_pool.go
Comment (RamLavi) [inline pkg/pool-manager/virtualmachine_pool.go:434]: ditto, I would move to the preliminary commit

**Review by RamLavi — CHANGES_REQUESTED** [review]
Hey @fossedihelm, thanks for your contribution!
Please see comments added.

**Comment by RamLavi** [comment]
/hold

let's merge only after stable branches are introduced to KMP

**Comment by fossedihelm** [comment]
Thanks @RamLavi for the review!
I am working on the relevant failing e2e tests. Thank you!

File: pkg/pool-manager/macentry_test.go
Comment (fossedihelm) [reply pkg/pool-manager/macentry_test.go:22]: Done!

File: pkg/pool-manager/pod_pool.go
Comment (fossedihelm) [reply pkg/pool-manager/pod_pool.go:307]: Done!

File: pkg/pool-manager/pool.go
Comment (fossedihelm) [reply pkg/pool-manager/pool.go]: Nice idea! Thanks! Done

**Comment by fossedihelm** [comment]
[Changes](https://github.com/k8snetworkplumbingwg/kubemacpool/compare/2479b07cc20aa2c5f3347e5fc9d413852b6f2510..447fd6637834eb509bf739eb7f398a004641f8f5):

- Addressed comments
- Fixed failing e2e test:
--  there were a mixed comparison between `iface.MacAddress` and `macKey.String()`
-- the first one was upper case, the latter in lower case.
- Added `MarshalJSON() ([]byte, error)`  to macMap struct to allow logger to correctly print the map

File: pkg/pool-manager/macpoolmap.go
Comment (RamLavi) [inline pkg/pool-manager/macpoolmap.go:101]: can we use the `String` method?
```suggestion
		if NewMacKey(iface.MacAddress).String() == macAddress {
```

I think it works, see [here](https://go.dev/play/p/c0PoTSrbmF1)

File: pkg/pool-manager/mackey.go
Comment (RamLavi) [inline pkg/pool-manager/mackey.go:20]: Since we are not changing the value of the object, there is no need to use a pointer to the object
```suggestion
func (m macKey) String() string {
```

This has an added value of being able to call the `String()` right after creating the object: `NewMacKey(MacAddress).String()`

**Review by RamLavi — CHANGES_REQUESTED** [review]
thanks for the changes! 
Big like for the Marshal function added.

Added one more comment.

File: pkg/pool-manager/virtualmachine_pool.go
Comment (RamLavi) [inline pkg/pool-manager/virtualmachine_pool.go:531]:
```suggestion
		if iface.MacAddress != "" && NewMacKey(iface.MacAddress).String() == macAddress {
```

**Review by RamLavi — CHANGES_REQUESTED** [review]
last one :)

**Review by RamLavi — APPROVED** [review]
/lgtm
/approve

@fossedihelm does this change need to be backported? 
If not, then I'll remove the old after the CVE storm passes and I add the appropriate stable branches...

**Pull request merged**
